infra/Pos/Diffusion/Subscription/Common.hs

```diff
@@ -96,6 +117,18 @@ subscribeTo keepAliveTimer sendActions peer = do
     msgSubscriptionTerminated :: NodeId -> SubscriptionTerminationReason -> Text
     msgSubscriptionTerminated = sformat $ "subscriptionWorker: lost connection to "%shown%" "%shown
 
+    timeSpecToMilliseconds :: TimeSpec -> Millisecond
+    timeSpecToMilliseconds = fromMicroseconds . div 1000 . toNanoSecs
+
+    alterPeerSubStatus :: Maybe SubscriptionStatus -> m ()
+    alterPeerSubStatus s = atomically $ do
+        stats <- readTVar subStatus
+        let !stats' = Map.alter fn peer stats
+        writeTVar subStatus stats'
+        where
+            fn :: Maybe SubscriptionStatus -> Maybe SubscriptionStatus
+            fn x = getOption (Option x <> Option s)
```

Comment:
Nice.

`alterPeerSubStatus Nothing` should go in a `finally` on `subscribeTo` I think, so that if the thing is killed even by an async exception we always bring the state back to consistency.